crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,394 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO(william) add metrics
+// TODO(william) make configurable
+
+use either::Either;
+use futures::future::join_all;
+use futures::stream::FuturesOrdered;
+
+use sui_macros::nondeterministic;
+use sui_types::base_types::{ObjectDigest, ObjectID};
+use tracing::debug;
+use typed_store::Map;
+
+use std::path::Path;
+use std::sync::Arc;
+
+use fastcrypto::hash::MultisetHash;
+use mysten_metrics::spawn_monitored_task;
+use sui_types::accumulator::Accumulator;
+use sui_types::committee::EpochId;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::TransactionEffects;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+use tokio_stream::StreamExt;
+use typed_store::rocks::TypedStoreError;
+use typed_store::rocks::{DBMap, MetricConf};
+use typed_store::traits::TableSummary;
+use typed_store::traits::TypedStoreDebug;
+use typed_store_derive::DBMapUtils;
+
+use crate::authority::authority_notify_read::NotifyRead;
+
+type AccumulatorTaskBuffer = FuturesOrdered<JoinHandle<SuiResult<(Accumulator, State)>>>;
+type EndOfEpochFlag = bool;
+
+#[derive(Debug, Clone)]
+pub struct State {
+    pub effects: Vec<TransactionEffects>,
+    pub checkpoint_seq_num: CheckpointSequenceNumber,
+    pub epoch: EpochId,
+    pub end_of_epoch_flag: EndOfEpochFlag,
+}
+
+#[derive(DBMapUtils)]
+pub struct StateAccumulatorTables {
+    // TODO: implement pruning policy as most of these tables can
+    // be cleaned up at end of epoch
+
+    // Maps checkpoint sequence number to an accumulator with accumulated state
+    // only for the checkpoint that the key references. Append-only, i.e.,
+    // the accumulator is complete wrt the checkpoint
+    pub state_hash_by_checkpoint: DBMap<CheckpointSequenceNumber, Accumulator>,
+
+    // A live, append-only table representing the running root state hash, computed at
+    // arbitrary checkpoints. As such, should NEVER be relied upon for snapshot
+    // verification or consensus.For key with Checkpoint C, the value is the accumulation
+    // of all state from checkpoint 0..C
+    pub(self) root_state_hash_by_checkpoint: DBMap<CheckpointSequenceNumber, Accumulator>,
+
+    // Finalized root state accumulator for epoch, to be included in CheckpointSummary
+    // of last checkpoint of epoch. These values should only ever be written once
+    // and never changed
+    pub root_state_hash_by_epoch: DBMap<EpochId, Accumulator>,
```

Comment:
do we still need this table, or can we instead enforce that `root_state_hash_by_checkpoint` will always have an entry for the final checkpoint of each epoch?

(this is not a big deal, just wondering if it would be simpler. it may not be simpler since it would make epoch state hash lookups indirect)

Reply:
The only real reason for my keeping this around is that we don't know the last checkpoint of the epoch in general. The only callsite, `CheckpointBuilder`, does though. I'm ok with removing

Reply:
I take that back. I think it's worth leaving around as it makes pruning easy - just save all of `root_state_hahs_by_epoch` , and in effect, we can always prune everything but the last key in `root_state_hash_by_checkpoint` without worrying too much

Reply:
ok agreed - but it seems like you maybe mistyped in your previous comment? I think that:
- `root_state_hash_by_epoch` goes in the perpetual store and doesn't need to be pruned (maybe sometime around epoch 1 million we can talk about pruning it)
- everything else goes in the per-epoch store so we get pruning for free?